### PR TITLE
WIP separate directories

### PR DESCRIPTION
**Pull request by JohannesGaessler**
Fixes https://github.com/hlky/stable-diffusion-webui/issues/23 .

Current state of the implementation of separate directories for saving samples.
I have solved the problem of prompt collision by simply logging file names and associated prompts to a text file.
This also helps in cases where people don't remember the full prompt.

I've had an idea for an alternative, however:
You could simply write a YAML file that encapsulates the entire GUI settings for each generated sample.
This would enable people to reliably reconstruct or share the setup for the images they previously generated.
With the script support that I implemented earlier today those YAML files could already be used.
For the GUI it would probably make sense to implement something like a load button (I don't care though).
Implementation would not be difficult.

Thoughts?

Also: there seems to be a branch called "23-separate-directories-for-samples".
Is there someone else working on this?

### CODE REVIEW

File: webui.py
Comment (hlky) [inline webui.py:413]: is this meant to be the same?

**Review by hlky — COMMENTED** [review]
seems fine so far, i assume you will be adding the gui code and txt2img/img2img code soon

File: webui.py
Comment (JohannesGaessler) [reply webui.py:413]: I didn't rigorously check the code since I wanted to confirm with you regarding the goals first ("bugs" during the specification phase always take the most effort to fix afterwards).

The implementation is supposed to be equivalent if `flat_outdir` is true.

**Comment by JohannesGaessler** [comment]
I'm assuming you agree with writing info files for the individual samples then.
I will implement code that enables a toggle of `flat_outdir` via GUI.
I may implement code for loading YAML files in the GUI depending on my mood and how tedious it is.

**Comment by hlky** [comment]
yeah info is good, whatever you think is good is fine, if it doesn't end up working like someone else wants they can create an issue with what they think is wrong

**Comment by JohannesGaessler** [comment]
I mean, it's your repo so you can do whatever you want but in my experience changing things after the fact is always the worst part of any project.
Making sure everyone is on the same page minimizes work in the long run.

File: webui.py
Comment (JohannesGaessler) [reply webui.py:413]: Ah, when I continued editing I understood what you meant.
This is a bug.
I misinterpreted what you were talking about because the conversation page only shows the lines _above_ your comment.

**Comment by JohannesGaessler** [comment]
Implementation is rather ugly due to the toggles.
Their order depends on whether `txt2img` or `img2img` is used.
Also this required passing more arguments to `process_images`; the function is getting pretty bloated.

Known problems:

When I try to use img2img I get the following exception:
```
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/gradio/routes.py", line 247, in run_predict
    output = await app.blocks.process_api(
  File "/usr/lib/python3.10/site-packages/gradio/blocks.py", line 639, in process_api
    processed_input = self.preprocess_data(fn_index, raw_input, state)
  File "/usr/lib/python3.10/site-packages/gradio/blocks.py", line 543, in preprocess_data
    processed_input.append(block.preprocess(raw_input[i]))
  File "/usr/lib/python3.10/site-packages/gradio/components.py", line 1563, in preprocess
    mask_im = processing_utils.decode_base64_to_image(mask)
  File "/usr/lib/python3.10/site-packages/gradio/processing_utils.py", line 28, in decode_base64_to_image
    content = encoding.split(";")[1]
AttributeError: 'NoneType' object has no attribute 'split'
```
Since this is entirely outside the project directory I don't understand what is the problem.
When I toggle and untoggle the buttons it suddenly works.

File indices get incremented by 2 instead of 1 because of the extra files.
Shouldn't be difficult to fix with regex/glob though.

It's already late where I live and I need to stop now or I won't be able to sleep.
You can either wait for me to come back or try to finish things up yourself; your call.

**Comment by JohannesGaessler** [comment]
I was able to reproduce the bug with `img2img` with the commit this branch is based on.
This suggests to me that I simply found a bug unrelated to my changes.

I fixed the numbering of png files using a glob pattern.
This is a UNIX style pattern but Python should interpret this correctly even on Windows.

Do you want to resolve the conflicts or should I do it?

File: webui.py
Comment (hlky) [inline webui.py:754]: img2img toggles rn

```
prompt_matrix = 0 in toggles
    normalize_prompt_weights = 1 in toggles
    loopback = 2 in toggles
    random_seed_loopback = 3 in toggles
    skip_save = 4 not in toggles
    skip_grid = 5 not in toggles
    use_GFPGAN = 6 in toggles
    use_RealESRGAN = 7 in toggles
```

File: webui.py
Comment (hlky) [inline webui.py:631]: also needs fixing, txt2img toggles

```
prompt_matrix = 0 in toggles
    normalize_prompt_weights = 1 in toggles
    skip_save = 2 not in toggles
    skip_grid = 3 not in toggles
    use_GFPGAN = 4 in toggles
    use_RealESRGAN = 5 in toggles
```

**Review by hlky — CHANGES_REQUESTED** [review]
Fix please

**Comment by hlky** [comment]
Can you just forcepush/update your branch? should help to resolve conflicts

**Comment by JohannesGaessler** [comment]
This already is the newest version; force-pushing won't do anything.
The problem is just that my changes conflict with changes made on master since yesterday, when I started working on this.
The solution is to either rebase this branch or to merge the branches.

I've just had lunch and coffee, expect a rebased version soon.

**Comment by hlky** [comment]
> rebase this branch or to merge the branches.

This is what I meant, sorry, too many words in the github vocabulary 

Thanks for your effort though, good work!

**Comment by hlky** [comment]
all done?

**Review by hlky — APPROVED** [review]
seems good

**Comment by JohannesGaessler** [comment]
I rebased onto master.
In my testing the features I added work correctly for txt2img.
I was **not** able to test my changes for img2img because it is broken for me on master.
I did **not** test if I accidentally broke settings because I don't have a good understanding of the intended implementation.
I also did **not** test if I accidentally broke Real-ESRGAN because I do not have it set up.

I did not test whether the changes on master broke CLI usage but I'll just make a new PR if there is a problem.

I unified the documentation of the YAML files.

This is now ready to merge from my side.

More general comments:
I would suggest that you rebase future PRs onto master before merging them.
It's essentially the same effort as just merging directly but the git log is way easier to understand.
Also makes it easier to identify bad commits with git blame.

**Comment by JohannesGaessler** [comment]
tfw just two seconds too late

**Comment by hlky** [comment]
You did actually test it, right?

```
glob() got an unexpected keyword argument 'root_dir'
```

**Comment by hlky** [comment]
[ec43d18](https://github.com/hlky/stable-diffusion-webui/commit/ec43d18ee533f5e81599050e2596e024be8a583d) fixed

**Comment by JohannesGaessler** [comment]
Regarding glob: 
I forgot that the `root_dir` argument was only added with Python 3.10, sorry.

Regarding img2img: 
I rebased to the best of my ability but my understanding of the img2img code is limited.
I only noticed that you had already partially merged the conflicts after the fact because I only saw your comments, not the line a little further up.

Regarding defaults:
I explicitly asked about this twice in the related issue and got no reply from you.
I interpreted this as you leaving the decision up to me (which I also explicitly asked about).

I do not understand what you mean by "original filename used".

Honestly this is why communication is important.
When there is a "bug" in the specification people implement the wrong things and time is wasted fixing things after the fact.

**Comment by JohannesGaessler** [comment]
Regarding the change in filenames:
This is also something I explicitly asked about in the original issue and got no reply on.
I think doing it like this is better.

**Pull request merged**
